src/chips/sam4l/flashcalw.rs

```diff
@@ -0,0 +1,1103 @@
+    while !flash.get_ready_status() { 
+        panic!("Flash wasn't ready for a command -- interrupt system used incorrectly.");
+    }
+}
+
+/// Trait for a client of the flash controller.
+pub trait Client {
+    //  Called upon a completed call
+    fn command_complete(&self, err: Error);     
+}
+    
+impl FLASHCALW {
+
+    pub fn mark_ready(&self) {
+        self.ready.put(Some(true));
+    }
+
+
+    pub fn get_ready_status(&self) -> bool {
+        if self.ready.is_none() || !self.ready.take().unwrap() {
+            self.ready.put(Some(false));
+            false
+        } else {
+            self.ready.put(Some(true));
+            true
+        }
+    }
+    
+    const fn new(base_addr: *mut Registers, ahb_clk: pm::HSBClock,
+    hramc1_clk: pm::HSBClock, pb_clk: pm::PBBClock) -> FLASHCALW {
+        FLASHCALW {
+            registers:              base_addr,
+            ahb_clock:              pm::Clock::HSB(ahb_clk),
+            hramc1_clock:           pm::Clock::HSB(hramc1_clk),
+            pb_clock:               pm::Clock::PBB(pb_clk),
+            wait_until_ready:       default_wait_until_ready,
+            error_status:           TakeCell::new(0),
+            ready:                  TakeCell::new(true),
+            client:                 TakeCell::empty(),
+            current_state:          Cell::new(FlashState::Unconfigured),
+            current_command:        Cell::new(Command::None),
+            page:                   Cell::new(0),
+            page_buffer:            TakeCell::new([0; FLASH_PAGE_SIZE as usize])
+        }
+    }
+
+    // Helper to read a flashcalw register 
+    fn read_register(&self, key : RegKey) -> u32 {
+        let registers : &mut Registers = unsafe { 
+            mem::transmute(self.registers)
+        };
+        
+        match key {
+            RegKey::CONTROL => {
+                registers.control.get()
+            },
+            RegKey::COMMAND => {
+                registers.command.get()
+            },
+            RegKey::STATUS => {
+                registers.status.get()
+            },
+            RegKey::PARAMETER => {
+                registers.parameter.get()
+            },
+            RegKey::VERSION => {
+                registers.version.get()
+            },
+            RegKey::GPFRHI => {
+                registers.general_purpose_fuse_register_hi.get()
+            },
+            RegKey::GPFRLO => {
+                registers.general_purpose_fuse_register_lo.get()
+            } 
+        }
+    }
+
+
+    pub fn handle_interrupt(&self) {
+        unsafe { 
+            //  mark the controller as ready and clear pending interrupt 
+            self.mark_ready();  
+            nvic::clear_pending(nvic::NvicIdx::HFLASHC);
+        }
+       
+        self.error_status.put(Some(self.get_error_status()));
+        let error_status = self.error_status.take().unwrap();
+        self.error_status.put(Some(error_status));
```

Comment:
Just reorganized.
```
let error_status= self.get_error_status();
self.error_status.put(Some(error_status));
```
The local error_status is used a few lines below to tell the client if there was an error, what error it was.